viadot/tasks/cloud_for_customers.py

```diff
@@ -1,46 +1,133 @@
-from prefect import task
+from prefect import task, Task
 import pandas as pd
 from ..sources import CloudForCustomers
 from typing import Any, Dict, List
+from prefect.utilities.tasks import defaults_from_attrs
 
 
-@task
-def c4c_report_to_df(report_url: str, env: str = "QA", skip: int = 0, top: int = 1000):
-    final_df = pd.DataFrame()
-    next_batch = True
-    while next_batch:
-        new_url = f"{report_url}&$top={top}&$skip={skip}"
-        chunk_from_url = CloudForCustomers(report_url=new_url, env=env)
-        df = chunk_from_url.to_df()
-        final_df = final_df.append(df)
-        if not final_df.empty:
-            df_count = df.count()[1]
-            if df_count != top:
-                next_batch = False
-            skip += top
-        else:
-            break
-    return final_df
-
-
-@task
-def c4c_to_df(
-    url: str = None,
-    endpoint: str = None,
-    report_url: str = None,
-    fields: List[str] = None,
-    params: Dict[str, Any] = {},
-    env: str = "QA",
-    if_empty: str = "warn",
-):
-    cloud_for_customers = CloudForCustomers(
-        url=url,
-        report_url=report_url,
-        endpoint=endpoint,
-        params=params,
-        env=env,
+class C4CReportToDF(Task):
+    def __init__(
+        self,
+        *args,
+        report_url: str = None,
+        env: str = "QA",
+        skip: int = 0,
+        top: int = 1000,
+        **kwargs,
+    ):
+
+        self.report_url = report_url
+        self.env = env
+        self.skip = skip
+        self.top = top
+
+        super().__init__(
+            name="c4c_report_to_df",
+            *args,
+            **kwargs,
+        )
+
+    def __call__(self, *args, **kwargs):
+        """Download report to DF"""
+        return super().__call__(*args, **kwargs)
+
+    @defaults_from_attrs(
+        "report_url",
+        "env",
+        "skip",
+        "top",
     )
+    def run(
+        self,
+        report_url: str = None,
+        env: str = "QA",
+        skip: int = 0,
+        top: int = 1000,
+    ):
+        """
+        Task for downloading data from the Cloud for Customers to a pandas DataFrame using report URL
+        (generated in Azure Data Factory).
```

Comment:
Can you explain the difference between report and non-report here or point to C4C documentation explaining it?

Reply:
added